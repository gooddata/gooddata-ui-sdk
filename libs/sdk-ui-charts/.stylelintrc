{
  "extends": [
    "../../.stylelintrc"
  ],
  "rules": {
    "max-nesting-depth": 4
  }
}
