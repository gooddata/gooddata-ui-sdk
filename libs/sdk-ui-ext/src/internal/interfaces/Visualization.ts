// (C) 2019-2021 GoodData Corporation
import { ISeparators } from "@gooddata/numberjs";
import isEmpty from "lodash/isEmpty";
import { IAnalyticalBackend, IExecutionFactory, ISettings, ITheme } from "@gooddata/sdk-backend-spi";
import {
    IColorPalette,
    Identifier,
    IInsight,
    IInsightDefinition,
    ITotal,
    ObjRef,
    VisualizationProperties,
    IExecutionConfig,
} from "@gooddata/sdk-model";
import {
    ChartType,
    ErrorCodes,
    GoodDataSdkError,
    IDrillableItem,
    IDrillEvent,
    IHeaderPredicate,
    ILocale,
    IPushData,
    IVisualizationCallbacks,
    OverTimeComparisonType,
    SdkErrorType,
    VisualizationEnvironment,
} from "@gooddata/sdk-ui";

export type RenderFunction = (component: any, target: Element) => void;

export interface IVisConstruct {
    backend: IAnalyticalBackend;
    // TODO: rename to workspace
    projectId: string;
    element: string;
    configPanelElement: string;

    callbacks: IVisCallbacks;
    environment?: VisualizationEnvironment;
    locale?: ILocale;
    // TODO: rename to settings
    featureFlags?: ISettings;
    visualizationProperties: VisualizationProperties;
    renderFun: RenderFunction;
}

export interface ICustomProps {
    drillableItems?: Array<IDrillableItem | IHeaderPredicate>;
    totalsEditAllowed?: boolean;
}

export interface IDimensions {
    width?: number; // Note: will not be optional once we start sending it
    height: number;
}

/**
 * @alpha
 */
export interface IVisProps {
    dimensions?: IDimensions;
    custom?: ICustomProps;
    locale?: ILocale;
    dateFormat?: string;
    config?: IGdcConfig;
    executionConfig?: IExecutionConfig;
    theme?: ITheme;

    /*
     * This can be used to override configuration of visualization when embedding through InsightView. The pluggable
     * visualization implementation MUST take these custom properties and override the the effective visualization
     * config with them.
     *
     * It is important that the override happens just before sending the config to the underlying chart -
     * the custom config passed through here will be in the format that is understood by the underlying chart.
     */
    customVisualizationConfig?: any;
}

export interface IVisualizationOptions {
    dateOptionsDisabled: boolean;
}

export interface IVisCallbacks extends IVisualizationCallbacks {
    pushData(data: IPushData, options?: IVisualizationOptions): void;
}

export interface IBucketFilterElement {
    title: string;
    uri: string;
}

export type ComparisonConditionOperator =
    | "GREATER_THAN"
    | "GREATER_THAN_OR_EQUAL_TO"
    | "LESS_THAN"
    | "LESS_THAN_OR_EQUAL_TO"
    | "EQUAL_TO"
    | "NOT_EQUAL_TO";

export interface IComparisonCondition {
    readonly comparison: {
        readonly operator: ComparisonConditionOperator;
        readonly value: number;
    };
}

export type RangeConditionOperator = "BETWEEN" | "NOT_BETWEEN";

export interface IRangeCondition {
    readonly range: {
        readonly operator: RangeConditionOperator;
        readonly from: number;
        readonly to: number;
    };
}

export type IMeasureValueFilterCondition = IComparisonCondition | IRangeCondition;

export interface IBucketFilterInterval {
    granularity: string;
    interval: string[];
    name: string;
}

export interface IAttributeFilter {
    attribute: string;
    isInverted: boolean;
    totalElementsCount: number;
    displayFormRef: ObjRef;
    selectedElements: Array<{
        title: string;
        uri: string;
    }>;
}

export const DATE_DATASET_ATTRIBUTE = "attr.datedataset";

export interface IDateFilter {
    attribute: "attr.datedataset";
    overTimeComparisonType: OverTimeComparisonType;
    interval: {
        granularity: string;
        interval: [string, string] | [number, number];
        name: string;
        type: "relative" | "absolute";
    };
}

export interface IMeasureValueFilter {
    measureLocalIdentifier: string;
    condition?: IMeasureValueFilterCondition;
}

export type RankingFilterOperator = "TOP" | "BOTTOM";

export interface IRankingFilter {
    measure: string;
    attributes?: string[];
    operator: RankingFilterOperator;
    value: number;
}

export type IBucketFilter = IAttributeFilter | IDateFilter | IMeasureValueFilter | IRankingFilter;

export interface ISort {
    direction: "asc" | "desc";
}

export interface IBucketItem {
    localIdentifier: string;
    type?: string;
    aggregation?: boolean;
    attribute?: string;
    filters?: IBucketFilter[];
    format?: string;
    granularity?: string;
    showInPercent?: boolean;
    showOnSecondaryAxis?: boolean;
    sort?: ISort;
    masterLocalIdentifier?: string;
    overTimeComparisonType?: OverTimeComparisonType;
    operandLocalIdentifiers?: Array<string | null> | null;
    operator?: string | null;

    dfRef?: ObjRef;
    locationDisplayFormRef?: ObjRef;
    dateDatasetRef?: ObjRef;
}

export interface IFiltersBucketItem extends IBucketItem {
    autoCreated?: boolean;
}

// TODO: SDK8: rename this :) the original name IBucket conflicted with what we have in model;
// this interface and all the other similar interface (bucket item, filters etc) are specifically used
// in the reference point
export interface IBucketOfFun {
    localIdentifier: string;
    items: IBucketItem[];
    totals?: ITotal[];
    chartType?: string;
}

export interface IFilters {
    localIdentifier: "filters";
    items: IFiltersBucketItem[];
}

export interface IRecommendations {
    [key: string]: boolean;
}

export interface IBucketUiConfig {
    accepts?: string[];
    canAddItems?: boolean;
    warningMessage?: string;
    title?: string;
    subtitle?: string;
    icon?: string;
    allowsDuplicateItems?: boolean;

    // allow more than one date item in the bucket, regardless of date dimension
    allowsDuplicateDates?: boolean;
    allowsReordering?: boolean;
    allowsSwapping?: boolean;
    enabled?: boolean;
    itemsLimit?: number;
    itemsLimitByType?: {
        date?: number;
        metric?: number;
        fact?: number;
        attribute?: number;
    };
    isShowInPercentEnabled?: boolean;
    isShowInPercentVisible?: boolean;
    isShowOnSecondaryAxisVisible?: boolean;
    allowShowOnSecondaryAxis?: boolean;
    allowSelectChartType?: boolean;
    allowOptionalStacking?: boolean;

    // indicates that the visualization prefers date items in the bucket to have the same dimension
    preferSynchronizedDates?: boolean;
}

export interface IBucketsUiConfig {
    [localIdentifier: string]: IBucketUiConfig;
}

export interface IExportUiConfig {
    supported?: boolean;
}

export interface INoMetricUiConfig {
    supported?: boolean;
}

export interface IOpenAsReportUiConfig {
    supported?: boolean;
    warningMessage?: string;
}

export interface ICustomError {
    heading: string;
    text: string;
}

export interface IOptionalStacking {
    supported?: boolean;
    disabled?: boolean;
    stackMeasures?: boolean;
    stackMeasuresToPercent?: boolean;
    canStackInPercent?: boolean;
}

export interface ISupportedLocationIcon {
    supported?: boolean;
}

export interface IUiConfig {
    buckets: IBucketsUiConfig;
    recommendations?: IRecommendations;
    exportConfig?: IExportUiConfig;
    noMetricAccepted?: INoMetricUiConfig;
    openAsReport?: IOpenAsReportUiConfig;
    customError?: ICustomError;
    supportedOverTimeComparisonTypes?: OverTimeComparisonType[];
    supportedChartTypes?: ChartType[];
    axis?: string;
    optionalStacking?: IOptionalStacking;
    supportedLocationIcon?: ISupportedLocationIcon;
}

export interface IVisualizationProperties {
    // This can be anything depending on a visualization type
    // perhaps consider adding: sortItems?: AFM.SortItem[]
    [property: string]: any;
}

export interface IReferencePoint {
    buckets: IBucketOfFun[];
    filters: IFilters;
    properties?: IVisualizationProperties; // properties are under plugvis creator's control
}

export interface IReferences {
    [identifier: string]: string;
}

export interface IExtendedReferencePoint {
    buckets: IBucketOfFun[];
    filters: IFilters;
    properties?: IVisualizationProperties; // properties are under plugvis creator's control
    uiConfig: IUiConfig;
}

/**
 * @alpha
 */
export interface IVisualization {
    /**
     * Update and re-render visualization to reflect change of insight state.
     *
     * Currently it is possible that context (AD/KD) will send insight that is not in a valid state:
     *
     * -  insight might be empty
     * -  insight might not be completely defined (user did not yet specify measures etc)
     *
     * It is the responsibility of the implementation to verify the state of insight during the update and
     * if there is anything amiss communicate with the context using the onError callback which it
     * received during construction time via {@link IVisConstruct}
     *
     * The loading state of the visualization must be communicated using the onLoadingChanged callback which
     * is also passed during construction time.
     *
     * @param props - some runtime properties
     * @param insight - new state of insight
     * @param insightPropertiesMeta - new state of insight properties meta
     * @param executionFactory - execution factory to use when triggering calculation on backend
     */
    update(
        props: IVisProps,
        insight: IInsightDefinition,
        insightPropertiesMeta: any,
        executionFactory: IExecutionFactory,
    ): void;

    unmount(): void;

    addNewDerivedBucketItems(
        referencePoint: IReferencePoint,
        newDerivedBucketItems: IBucketItem[],
    ): Promise<IReferencePoint>;

    /**
     * Called every time the reference point or the visualization class change
     * to allow visualizations to get updated ExtendedReferencePoint.
     * @param referencePoint The new value of the reference point.
     * @param previousReferencePoint The previous value of the reference point.
     * This value is only provided if the visualization class was not changed
     * (i. e. both points are related to the same visualization class).
     * @returns Promise of the new ExtendedReferencePoint.
     */
    getExtendedReferencePoint(
        referencePoint: IReferencePoint,
        previousReferencePoint?: IReferencePoint,
    ): Promise<IExtendedReferencePoint>;

    /**
     * Called when the Drill Down is performed, used to get the Drill Down target {@link IInsight} instance.
     *
     * The exact contract depends on individual {@link IInsight} type, but generally it should replace
     * the drilled attribute with the Drill Down target target attribute and include the filters from the
     * drill event into the returned {@link IInsight}.
     *
     * @param source {@link IInsight} to be used for the the Drill Down execution
     * @param drillDownContext Drill Down configuration used to properly create the result
     * @returns {@code source} as the Drill Down target {@link IInsight}
     */
    getInsightWithDrillDownApplied(source: IInsight, drillDownContext: IDrillDownContext): IInsight;
}

export interface IGdcConfig {
    separators?: ISeparators;
    colorPalette?: IColorPalette;
    isExportMode?: boolean;
    isInEditMode?: boolean;
    mapboxToken?: string;
    maxWidth?: number;
    maxHeight?: number;
    forceDisableDrillOnAxes?: boolean;
}

/**
 * Class name of element where pluggable visualization is supposed to render its configuration
 * panels.
 *
 * @alpha
 */
export const ConfigPanelClassName = "gd-configuration-panel-content";

/**
 * @alpha
 */
export const PluggableVisualizationErrorCodes = {
    /**
     * If pluggable visualization is asked to render itself but its buckets do not contain the right 'stuff',
     * then this is the error code to communicate the fact.
     */
    INVALID_BUCKETS: "INVALID_BUCKETS",

    /**
     * This error means that empty AFM was went to the GoodData.UI and as such can't be executed.
     */
    EMPTY_AFM: "EMPTY_AFM",
};

/**
 * @alpha
 */
export type PluggableVisualizationErrorType = keyof typeof PluggableVisualizationErrorCodes;

/**
 * @alpha
 */
export class InvalidBucketsSdkError extends GoodDataSdkError {
    public readonly pveType: PluggableVisualizationErrorType;

    constructor(cause?: Error) {
        super(ErrorCodes.UNKNOWN_ERROR as SdkErrorType, undefined, cause);

        this.pveType = "INVALID_BUCKETS";
    }

    public getErrorCode(): string {
        return this.pveType;
    }
}

/**
 * @alpha
 */
export class EmptyAfmSdkError extends GoodDataSdkError {
    public readonly pveType: PluggableVisualizationErrorType;

    constructor(cause?: Error) {
        super(ErrorCodes.UNKNOWN_ERROR as SdkErrorType, undefined, cause);

        this.pveType = "EMPTY_AFM";
    }

    public getErrorCode(): string {
        return this.pveType;
    }
}

/**
 * @alpha
 */
export type PluggableVisualizationError = InvalidBucketsSdkError | EmptyAfmSdkError;

/**
 * @alpha
 */
export function isPluggableVisualizationError(obj: unknown): obj is PluggableVisualizationError {
    return !isEmpty(obj) && (obj as any).pveType !== undefined;
}

/**
 * @alpha
 */
export function isInvalidBuckets(obj: unknown): obj is InvalidBucketsSdkError {
    return !isEmpty(obj) && (obj as InvalidBucketsSdkError).pveType === "INVALID_BUCKETS";
}

/**
 * @alpha
 */
export function isEmptyAfm(obj: unknown): obj is EmptyAfmSdkError {
    return !isEmpty(obj) && (obj as EmptyAfmSdkError).pveType === "EMPTY_AFM";
}

/**
 * Source implicit drill down attribute local Identifier
 *
 * @alpha
 */
export interface IDrillFromAttribute {
    drillFromAttribute: {
        localIdentifier: Identifier;
    };
}

/**
 * Target implicit drill down attribute display form
 *
 * @alpha
 */
export interface IDrillToAttribute {
    drillToAttribute: {
        attributeDisplayForm: ObjRef;
    };
}

/**
 * Implicit drill down definition
 *
 * @alpha
 */
export interface IImplicitDrillDown {
    implicitDrillDown: {
        from: IDrillFromAttribute;
        drillDownStep: IDrillToAttribute;
    };
}

/**
 * Implicit drill down context
 *
 * @alpha
 */
export interface IDrillDownContext {
    drillDefinition: IImplicitDrillDown;
    event: IDrillEvent;
}
