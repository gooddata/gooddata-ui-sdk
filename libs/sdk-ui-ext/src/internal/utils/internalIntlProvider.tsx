// (C) 2019-2024 GoodData Corporation
import React, { useMemo } from "react";
import { IntlProvider, IntlShape, createIntl, IntlConfig } from "react-intl";
import {
    DefaultLocale,
    ILocale,
    pickCorrectWording,
    TranslationsCustomizationProvider,
} from "@gooddata/sdk-ui";

import { translations } from "./translations.js";
import { IWorkspaceSettings } from "@gooddata/sdk-backend-spi";
import { LRUCache } from "lru-cache";

const INTL_CACHE_SIZE = 20;
const INTL_CACHE_KEY = "messages";
const intlCache = new LRUCache<string, IntlConfig>({ max: INTL_CACHE_SIZE });

function getLocaleStrings(locale) {
    const strings = translations[locale];
    if (!strings) {
        console.warn(`Missing locale strings for ${locale}, using ${DefaultLocale}`);
        return translations[DefaultLocale];
    }

    return strings;
}

export function createInternalIntl(locale: ILocale = DefaultLocale): IntlShape {
    /**
     * Because of issues described in the ticket FET-855, we decided to use this workaround.
     * After the issues that are described in the ticket are solved or at least reduced,
     * this workaround can be removed.
     */
    const cachedIntlConfig = intlCache.get(INTL_CACHE_KEY);
    if (cachedIntlConfig?.locale === locale) {
        return createIntl(cachedIntlConfig);
    }
    const settings = window.gdSettings as IWorkspaceSettings;
    const strings = getLocaleStrings(locale);

    intlCache.set(INTL_CACHE_KEY, {
        locale,
        messages: pickCorrectWording(strings, settings),
    });
    return createIntl(intlCache.get(INTL_CACHE_KEY));
}

interface IInternalIntlWrapperProps {
    locale?: string;
    workspace?: string;
    children?: React.ReactNode;
}

export const InternalIntlWrapper: React.FC<IInternalIntlWrapperProps> = ({
    locale = DefaultLocale,
    children,
    workspace,
}) => {
    /**
     * Because of issues described in the ticket FET-855, we decided to use this workaround.
     * After the issues that are described in the ticket are solved or at least reduced,
     * this workaround can be removed.
     */
    const settings = window.gdSettings as IWorkspaceSettings;

    const strings = getLocaleStrings(locale);
    const messages = useMemo(() => pickCorrectWording(strings, settings), [strings, settings]);

    if (settings) {
        return (
            <IntlProvider locale={locale} messages={messages}>
                {children}
            </IntlProvider>
        );
    } else {
        return (
            <TranslationsCustomizationProvider
                translations={strings}
                workspace={workspace}
                render={(modifiedTranslations) => {
                    return (
                        <IntlProvider key={locale} locale={locale} messages={modifiedTranslations}>
                            {children}
                        </IntlProvider>
                    );
                }}
            />
        );
    }
};
