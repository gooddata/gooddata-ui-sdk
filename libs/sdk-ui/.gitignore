styles/css
styles/internal/css

src/base/localization/bundles/*.ts
