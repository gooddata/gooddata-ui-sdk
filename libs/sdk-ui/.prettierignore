#we ignore these files because they are generated after build
src/base/localization/bundles/*.ts
