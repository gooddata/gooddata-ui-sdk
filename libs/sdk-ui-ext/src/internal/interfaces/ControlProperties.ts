// (C) 2023 GoodData Corporation
import { IComparison } from "@gooddata/sdk-ui-charts";

export interface IDefaultControlProperties {
    // This can be anything depending on a visualization type
    [property: string]: any;
}

export interface IComparisonControlProperties {
    comparison?: IComparison;
}

export type HeadlineControlProperties = IComparisonControlProperties;
