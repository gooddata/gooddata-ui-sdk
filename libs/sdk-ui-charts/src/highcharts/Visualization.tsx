// (C) 2007-2018 GoodData Corporation
import { IDataView } from "@gooddata/sdk-backend-spi";
import * as invariant from "invariant";
import * as React from "react";
import { IDrillableItem, OnFiredDrillEvent, IHeaderPredicate } from "@gooddata/sdk-ui";
import ChartTransformation, { renderHighCharts } from "./chart/ChartTransformation";
import Highcharts from "./chart/highcharts/highchartsEntryPoint";
import { IChartConfig } from "./Config";

import { isChartSupported, stringifyChartTypes } from "./utils/common";
import isEqual = require("lodash/isEqual");
import isFunction = require("lodash/isFunction");
import noop = require("lodash/noop");
import omitBy = require("lodash/omitBy");

export interface IVisualizationProps {
    height: number;
    width: number;
    config: IChartConfig;
    numericSymbols?: string[];

    dataView: IDataView;
    drillableItems: Array<IDrillableItem | IHeaderPredicate>;
    locale?: string;

    onDrill?: OnFiredDrillEvent;
    afterRender?: () => void;
    onDataTooLarge(): void;
    onNegativeValues(): void;
    onLegendReady(): void;
    pushData?(data: any): void;
}

export class Visualization extends React.Component<IVisualizationProps> {
    public static defaultProps = {
        locale: "en-US",
        numericSymbols: [] as string[],
        onDrill: () => true,
        afterRender: noop,
    };

    constructor(props: IVisualizationProps) {
        super(props);

        this.setNumericSymbols(this.props);
    }

    public UNSAFE_componentWillReceiveProps(nextProps: IVisualizationProps) {
        this.setNumericSymbols(nextProps);
    }

    public shouldComponentUpdate(nextProps: IVisualizationProps) {
        return !isEqual(omitBy(this.props, isFunction), omitBy(nextProps, isFunction));
    }

    public setNumericSymbols(props: IVisualizationProps) {
        const { numericSymbols } = props;

        if (numericSymbols && numericSymbols.length) {
            Highcharts.setOptions({
                lang: {
                    numericSymbols,
                },
            });
        }
    }

    public render(): JSX.Element {
        const visType = this.props.config.type;

        if (isChartSupported(visType)) {
            const {
                height,
                width,
                config,
                dataView,
                drillableItems,
                onDrill,
                afterRender,
                onDataTooLarge,
                onNegativeValues,
                onLegendReady,
                locale,
                pushData,
            } = this.props;

            return (
                <ChartTransformation
                    height={height}
                    width={width}
                    config={config}
                    drillableItems={drillableItems}
                    locale={locale}
                    dataView={dataView}
                    afterRender={afterRender}
                    onDrill={onDrill}
                    onDataTooLarge={onDataTooLarge}
                    onNegativeValues={onNegativeValues}
                    onLegendReady={onLegendReady}
                    pushData={pushData}
                    renderer={renderHighCharts}
                />
            );
        }

        invariant(
            false,
            `Unknown visualization type: ${visType}. Supported visualization types: ${stringifyChartTypes()}`,
        );

        return null;
    }
}
