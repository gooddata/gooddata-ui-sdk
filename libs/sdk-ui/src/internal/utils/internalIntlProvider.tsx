// (C) 2019 GoodData Corporation
import * as React from "react";
import { IntlProvider, IntlShape, createIntl } from "react-intl";
import { translations } from "@gooddata/js-utils";
import { DEFAULT_LOCALE } from "../../base/constants/localization";

import * as enUS from "../translations/en-US.json";
import * as deDE from "../translations/de-DE.json";
import * as esES from "../translations/es-ES.json";
import * as frFR from "../translations/fr-FR.json";
import * as jaJP from "../translations/ja-JP.json";
import * as nlNL from "../translations/nl-NL.json";
import * as ptBR from "../translations/pt-BR.json";
import * as ptPT from "../translations/pt-PT.json";
import * as zhHans from "../translations/zh-Hans.json";
import { ILocale } from "../../base/interfaces/Locale";

export const messagesMap = {
    "en-US": translations.removeMetadata(enUS),
    "de-DE": deDE,
    "es-ES": esES,
    "fr-FR": frFR,
    "ja-JP": jaJP,
    "nl-NL": nlNL,
    "pt-BR": ptBR,
    "pt-PT": ptPT,
    "zh-Hans": zhHans,
};

export function createInternalIntl(locale: ILocale = DEFAULT_LOCALE): IntlShape {
    return createIntl({ locale, messages: messagesMap[locale] });
}

export interface IInternalIntlWrapperProps {
    locale?: string;
}

export class InternalIntlWrapper extends React.PureComponent<IInternalIntlWrapperProps> {
    public static defaultProps: IInternalIntlWrapperProps = {
        locale: DEFAULT_LOCALE,
    };
    public render() {
        const { locale } = this.props;
        return (
            <IntlProvider locale={locale} messages={messagesMap[locale]}>
                {this.props.children}
            </IntlProvider>
        );
    }
}
