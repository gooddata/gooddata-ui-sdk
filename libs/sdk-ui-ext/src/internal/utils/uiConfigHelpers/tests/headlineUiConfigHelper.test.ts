// (C) 2019-2020 GoodData Corporation
import { describe, expect, it } from "vitest";

import { BucketNames, DefaultLocale } from "@gooddata/sdk-ui";
import { IBucket, newMeasure, newPopMeasure, newPreviousPeriodMeasure } from "@gooddata/sdk-model";
import { CalculateAs, DEFAULT_COMPARISON_PALETTE, IChartConfig } from "@gooddata/sdk-ui-charts";

import { createInternalIntl } from "../../internalIntlProvider.js";
import {
    buildHeadlineVisualizationConfig,
    getComparisonDefaultCalculationType,
    getHeadlineSupportedProperties,
    getHeadlineUiConfig,
    isComparisonEnabled,
} from "../headlineUiConfigHelper.js";
import * as referencePointMocks from "../../../tests/mocks/referencePointMocks.js";
import { IVisualizationProperties } from "../../../interfaces/Visualization.js";
import { HeadlineControlProperties } from "../../../interfaces/ControlProperties.js";
import { createTestProperties, newInsight } from "../../../tests/testDataProvider.js";

describe("headlineUiConfigHelper", () => {
    describe("getHeadlineUiConfig", () => {
        const intl = createInternalIntl(DefaultLocale);

        describe("'canAddItems' property", () => {
            const uiConfig = getHeadlineUiConfig(
                referencePointMocks.headlineWithMeasureInPrimaryBucket,
                intl,
            );

            it("should set 'canAddItems' bucket property falsy if it already contains a measure", () => {
                expect(uiConfig.buckets.secondary_measures.canAddItems).toBeTruthy();
            });

            it("should set 'canAddItems' bucket property truthy if it has no measures", () => {
                expect(uiConfig.buckets.measures.canAddItems).toBeFalsy();
            });
        });

        describe("'icon' property", () => {
            it("should set 'icon' property in both 'measures' and 'secondary_measures' buckets", () => {
                const uiConfig = getHeadlineUiConfig(
                    referencePointMocks.headlineWithMeasureInPrimaryBucket,
                    intl,
                );
                expect(uiConfig.buckets.measures.icon).toBeDefined();
                expect(uiConfig.buckets.secondary_measures.icon).toBeDefined();
            });
        });

        describe("'customError' property", () => {
            it("should set 'customError' property if there is a measure in 'secondary_measures' bucket, but 'measures' bucket is empty", () => {
                const uiConfig = getHeadlineUiConfig(
                    referencePointMocks.headlineWithMeasureInSecondaryBucket,
                    intl,
                );
                expect(uiConfig.customError).toHaveProperty("heading");
                expect(uiConfig.customError).toHaveProperty("text");
            });

            it("should keep 'customError' property empty if one of buckets contains a measure", () => {
                const uiConfig = getHeadlineUiConfig(
                    referencePointMocks.headlineWithMeasureInPrimaryBucket,
                    intl,
                );
                expect(uiConfig.customError).toBeUndefined();
            });
        });

        describe("measures bucket titles", () => {
            it("should set bucket titles", () => {
                const uiConfig = getHeadlineUiConfig(
                    referencePointMocks.headlineWithMeasureInPrimaryBucket,
                    intl,
                );
                expect(uiConfig.buckets.measures.title).toEqual("Measure");
                expect(uiConfig.buckets.secondary_measures.title).toEqual("Measure");
            });
        });
    });

    describe("isComparisonEnabled", () => {
        it("Should return true when bucket have 1 primary measure", () => {
            const buckets: IBucket[] = [
                {
                    localIdentifier: BucketNames.MEASURES,
                    items: [newMeasure("measure-1")],
                },
            ];

            const insight = newInsight(buckets);
            expect(isComparisonEnabled(insight)).toBe(false);
        });

        it("Should return true when bucket have 1 primary measure and 2 secondary measures", () => {
            const buckets: IBucket[] = [
                {
                    localIdentifier: BucketNames.MEASURES,
                    items: [newMeasure("measure-1")],
                },
                {
                    localIdentifier: BucketNames.SECONDARY_MEASURES,
                    items: [newMeasure("measure-2"), newMeasure("measure-3")],
                },
            ];

            const insight = newInsight(buckets);
            expect(isComparisonEnabled(insight)).toBe(false);
        });

        it("Should return false when bucket have 1 primary measure and 1 secondary measure", () => {
            const buckets: IBucket[] = [
                {
                    localIdentifier: BucketNames.MEASURES,
                    items: [newMeasure("measure-1")],
                },
                {
                    localIdentifier: BucketNames.SECONDARY_MEASURES,
                    items: [newMeasure("measure-2")],
                },
            ];

            const insight = newInsight(buckets);
            expect(isComparisonEnabled(insight)).toBe(true);
        });
    });

    describe("getHeadlineSupportedProperties", () => {
        it("Should return default control properties", () => {
            expect(getHeadlineSupportedProperties({})).toEqual(
                createTestProperties<HeadlineControlProperties>({
                    comparison: {
                        enabled: true,
                    },
                }),
            );
        });

        it("Should return properties override default properties", () => {
            const visualizationProperties = createTestProperties<HeadlineControlProperties>({
                comparison: {
                    enabled: false,
                    calculationType: CalculateAs.DIFFERENCE,
                },
            });
            const properties: IVisualizationProperties<HeadlineControlProperties> = {
                ...visualizationProperties,
                controls: {
                    ...visualizationProperties.controls,
                    color: ["red"],
                },
            };

            expect(getHeadlineSupportedProperties(properties)).toEqual(properties);
        });
    });

    describe("buildHeadlineVisualizationConfig", () => {
        it("Should build config correctly", () => {
            const properties = createTestProperties<HeadlineControlProperties>({
                comparison: {
                    enabled: false,
                    calculationType: CalculateAs.DIFFERENCE,
                },
            });

            const config: IChartConfig = {
                separators: {
                    thousand: ",",
                    decimal: ".",
                },
            };

            expect(buildHeadlineVisualizationConfig(properties, {}, { config })).toEqual({
                ...config,
                ...properties.controls,
                colorPalette: DEFAULT_COMPARISON_PALETTE,
            });
        });
    });

    describe("getComparisonDefaultCalculationType", () => {
        it("Should return calculation type as change in case secondary measure is pop measure", () => {
            const insight = newInsight([
                {
                    localIdentifier: BucketNames.SECONDARY_MEASURES,
                    items: [newPopMeasure("foo", "attr")],
                },
            ]);
            expect(getComparisonDefaultCalculationType(insight)).toEqual(CalculateAs.CHANGE);
        });

        it("Should return calculation type as change in case secondary measure is previous period measure", () => {
            const insight = newInsight([
                {
                    localIdentifier: BucketNames.SECONDARY_MEASURES,
                    items: [newPreviousPeriodMeasure("foo", [{ dataSet: "bar", periodsAgo: 3 }])],
                },
            ]);
            expect(getComparisonDefaultCalculationType(insight)).toEqual(CalculateAs.CHANGE);
        });

        it("Should return calculation type as ratio in case secondary measure is not a derived measure", () => {
            const insight = newInsight([
                {
                    localIdentifier: BucketNames.SECONDARY_MEASURES,
                    items: [newMeasure("foo")],
                },
            ]);
            expect(getComparisonDefaultCalculationType(insight)).toEqual(CalculateAs.RATIO);
        });
    });
});
