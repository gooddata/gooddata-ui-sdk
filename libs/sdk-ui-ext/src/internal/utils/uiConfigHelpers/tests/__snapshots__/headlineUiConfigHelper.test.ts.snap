// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`headlineUiConfigHelper > getComparisonColorPalette > should return a color palette with default colors when theme values are missing 1`] = `
[
  {
    "fill": {
      "b": 141,
      "g": 193,
      "r": 0,
    },
    "guid": "positive",
  },
  {
    "fill": {
      "b": 64,
      "g": 77,
      "r": 229,
    },
    "guid": "negative",
  },
  {
    "fill": {
      "b": 173,
      "g": 161,
      "r": 148,
    },
    "guid": "equals",
  },
]
`;

exports[`headlineUiConfigHelper > getComparisonColorPalette > should return a color palette with positive, negative, and equals colors 1`] = `
[
  {
    "fill": {
      "b": 0,
      "g": 255,
      "r": 0,
    },
    "guid": "positive",
  },
  {
    "fill": {
      "b": 0,
      "g": 0,
      "r": 255,
    },
    "guid": "negative",
  },
  {
    "fill": {
      "b": 255,
      "g": 0,
      "r": 0,
    },
    "guid": "equals",
  },
]
`;

exports[`headlineUiConfigHelper > getComparisonColorPalette > should return a color palette with positive, negative, and equals colors when kpi.secondaryInfoColor is empty 1`] = `
[
  {
    "fill": {
      "b": 0,
      "g": 255,
      "r": 0,
    },
    "guid": "positive",
  },
  {
    "fill": {
      "b": 0,
      "g": 0,
      "r": 255,
    },
    "guid": "negative",
  },
  {
    "fill": {
      "b": 204,
      "g": 204,
      "r": 204,
    },
    "guid": "equals",
  },
]
`;
